File: output.go
package main

import (
	"fmt"
	"html"
	"strings"
)

func stripAnsi(s string) string {
	// Strip CSI sequences: ESC [ ... letter
	var result strings.Builder
	i := 0
	for i < len(s) {
		if s[i] == 0x1B && i+1 < len(s) && s[i+1] == '[' {
			// Skip until a letter (终结符)
			j := i + 2
			for j < len(s) {
				c := s[j]
				if (c >= 'A' && c <= 'Z') || (c >= 'a' && c <= 'z') {
					i = j + 1
					break
				}
				j++
			}
			if j >= len(s) {
				break
			}
			continue
		}
		result.WriteByte(s[i])
		i++
	}
	return result.String()
}

func cleanOutput(raw string) string {
	cleaned := stripAnsi(raw)
	lines := strings.Split(cleaned, "\n")
	for i, line := range lines {
		lines[i] = strings.TrimRight(line, " \t\r")
	}
	return strings.TrimSpace(strings.Join(lines, "\n"))
}

func chunkMessage(text string, limit int) []string {
	if limit <= 0 {
		limit = 4000
	}
	if len(text) <= limit {
		return []string{text}
	}

	var chunks []string
	remaining := text

	for len(remaining) > 0 {
		if len(remaining) <= limit {
			chunks = append(chunks, remaining)
			break
		}

		cutAt := strings.LastIndex(remaining[:limit], "\n")
		if cutAt <= 0 {
			cutAt = limit
		}

		chunks = append(chunks, remaining[:cutAt])
		remaining = strings.TrimLeft(remaining[cutAt:], "\n")
	}

	return chunks
}

func FormatOutput(cmd, output, serverInfo string, execTimeMs int64) string {
	var sb strings.Builder

	if serverInfo != "" {
		sb.WriteString(fmt.Sprintf("<b>\U0001F5A5 %s</b>\n\n", html.EscapeString(serverInfo)))
	}

	sb.WriteString(fmt.Sprintf("<code>$ %s</code>\n\n", html.EscapeString(cmd)))

	if output == "" {
		sb.WriteString("<i>(no output)</i>")
	} else {
		sb.WriteString(fmt.Sprintf("<pre>%s</pre>", html.EscapeString(output)))
	}

	if execTimeMs >= 0 {
		sb.WriteString(fmt.Sprintf("\n\n<i>%dms</i>", execTimeMs))
	}

	return sb.String()
}

func FormatError(cmd, errMsg, serverInfo string) string {
	var sb strings.Builder

	if serverInfo != "" {
		sb.WriteString(fmt.Sprintf("<b>\U0001F5A5 %s</b>\n\n", html.EscapeString(serverInfo)))
	}

	sb.WriteString(fmt.Sprintf("<code>$ %s</code>\n\n", html.EscapeString(cmd)))
	sb.WriteString(fmt.Sprintf("<i>\u274C %s</i>", html.EscapeString(errMsg)))

	return sb.String()
}

File: src/utils/output.js
// Strip ANSI escape codes
export function stripAnsi(str) {
  return str
    .replace(/\x1B\[[0-9;]*[a-zA-Z]/g, '')
    .replace(/\x1B\][^\x07]*\x07/g, '')
    .replace(/\x1B[()][AB012]/g, '');
}

// Clean trailing whitespace from each line, trim overall
export function cleanOutput(raw) {
  return stripAnsi(raw)
    .split('\n')
    .map((l) => l.replace(/\r/g, '').replace(/\s+$/, ''))
    .join('\n')
    .trim();
}

// Split text into chunks under Telegram's 4096 char limit
export function chunkMessage(text, limit = 4000) {
  if (text.length <= limit) return [text];

  const chunks = [];
  let remaining = text;

  while (remaining.length > 0) {
    if (remaining.length <= limit) {
      chunks.push(remaining);
      break;
    }

    let cutAt = remaining.lastIndexOf('\n', limit);
    if (cutAt <= 0) cutAt = limit;

    chunks.push(remaining.substring(0, cutAt));
    remaining = remaining.substring(cutAt).replace(/^\n/, '');
  }

  return chunks;
}

function escHtml(s) {
  return s.replace(/&/g, '&amp;').replace(/</g, '&lt;').replace(/>/g, '&gt;');
}

// Format a successful command result
export function formatSuccess(cmd, output, serverInfo, execTime) {
  const timeStr = execTime !== null ? `${execTime}ms` : '';
  const header = serverInfo
    ? `<b>\u{1F5A5} ${escHtml(serverInfo)}</b>`
    : '';

  const cmdLine = `<code>$ ${escHtml(cmd)}</code>`;

  if (!output) {
    const parts = [header, cmdLine, '<i>(no output)</i>'];
    if (timeStr) parts.push(`<i>${timeStr}</i>`);
    return parts.filter(Boolean).join('\n\n');
  }

  const parts = [header, cmdLine, `<pre>${escHtml(output)}</pre>`];
  if (timeStr) parts.push(`<i>${timeStr}</i>`);
  return parts.join('\n\n');
}

// Format an error result
export function formatError(cmd, error, serverInfo) {
  const header = serverInfo
    ? `<b>\u{1F5A5} ${escHtml(serverInfo)}</b>`
    : '';

  const parts = [
    header,
    `<code>$ ${escHtml(cmd)}</code>`,
    `<i>\u274C ${escHtml(error)}</i>`,
  ];
  return parts.filter(Boolean).join('\n\n');
}

// Format connection status messages
export function formatConnected(host, port, username) {
  return [
    `\u2705 <b>Connected</b>`,
    `<code>${escHtml(username)}@${escHtml(host)}:${port}</code>`,
    '',
    `Send any message to execute commands.`,
    `Type /disconnect to end session.`,
  ].join('\n');
}

export function formatConnecting(host, port, username) {
  return `\u23F3 Connecting to <code>${escHtml(username)}@${escHtml(host)}:${port}</code>...`;
}

export function formatConnectionError(error) {
  return `\u274C Connection failed: <i>${escHtml(error)}</i>`;
}
